# illumos revision
--- usr/src/contrib/mDNSResponder/mDNSShared/Java/JNISupport.c	2017-07-24 09:23:31.000000000 -0500
+++ usr/src/lib/libdns_sd/java/common/JNISupport.c	2017-07-24 09:23:43.000000000 -0500
@@ -1069,4 +1069,8 @@ exit:
 
 // NOT static -- otherwise the compiler may optimize it out
 // The "@(#) " pattern is a special prefix the "what" command looks for
+#ifndef MDNS_VERSIONSTR_NODTS
 const char VersionString_SCCS[] = "@(#) libjdns_sd " STRINGIFY(mDNSResponderVersion) " (" __DATE__ " " __TIME__ ")";
+#else
+const char VersionString_SCCS[] = "@(#) libjdns_sd " STRINGIFY(mDNSResponderVersion);
+#endif
