# illumos revision
--- usr/src/contrib/mDNSResponder/mDNSShared/Java/DNSSDRecordRegistrar.java	2017-07-24 09:23:31.000000000 -0500
+++ usr/src/lib/libdns_sd/java/com/apple/dnssd/DNSSDRecordRegistrar.java	2017-07-24 09:23:39.000000000 -0500
@@ -43,15 +43,12 @@
 					The class of the resource record, as defined in nameser.h
 					(usually 1 for the Internet class).
 		<P>
-		@param	rData
+		@param	rdata
 					The new rdata  as it is to appear in the DNS record.
 		<P>
 		@param	ttl
 					The time to live of the resource record, in seconds. Pass 0 to use a default value.
 		<P>
-		@param	listener
-					This object will get called when the service is registered.
-		<P>
 		@return		A {@link DNSSDService} that can be used to abort the record registration.
 
 		@throws SecurityException If a security manager is present and denies <tt>RuntimePermission("getDNSSDInstance")</tt>.
