# illumos revision
--- usr/src/contrib/mDNSResponder/mDNSShared/dnssd_clientlib.c	2017-07-24 09:23:31.000000000 -0500
+++ usr/src/lib/libdns_sd/common/dnssd_clientlib.c	2017-07-24 09:23:36.000000000 -0500
@@ -363,4 +363,8 @@ DNSServiceErrorType DNSSD_API TXTRecordG
 
 // NOT static -- otherwise the compiler may optimize it out
 // The "@(#) " pattern is a special prefix the "what" command looks for
+#ifndef MDNS_VERSIONSTR_NODTS
 const char VersionString_SCCS_libdnssd[] = "@(#) libdns_sd " STRINGIFY(mDNSResponderVersion) " (" __DATE__ " " __TIME__ ")";
+#else
+const char VersionString_SCCS_libdnssd[] = "@(#) libdns_sd " STRINGIFY(mDNSResponderVersion);
+#endif
